### ISSUE TITLE
Log 500 exceptions with stack traces

### ISSUE THREAD

**Issue by sverhagen** [wimdeblauwe/error-handling-spring-boot-starter#29]
I see that `error.handling.exception-logging` lets you toggle between `WITH_STACKTRACE` and `MESSAGE_ONLY`. And that `full-stacktrace-classes` lets you affect the behavior. But the way I look at `500 Internal Server Error`s versus everything else is as follows:

* `500 Internal Server Error`: undefined behavior happened, the developer did not account for this, someone needs to evaluate this for future improvement. To support this troubleshooting, I would like these to be logged **with** stack traces
* **All other errors**: defined behavior, as undesirable as it may be, though often it's caused by bad requests from the client. To not overwhelm the log file, I would like these to be logged **without** stack traces

Unfortunately, the logging configuration described here is not currently possible. I would like a boolean property that would cause exceptions being logged for all 5xx or 500 status codes, and message-only for all others. Or a range, or something.

The fact that, to me, 500 status codes maps to undefined behavior also makes it undesirable to _define_ exceptions ahead of time in `full-stacktrace-classes`.

Thank you for this useful library!

**Comment by wimdeblauwe** [comment]
I implemented something that might work on branch `feature/log_stacktrace_http_status`, see https://github.com/wimdeblauwe/error-handling-spring-boot-starter/tree/feature/log_stacktrace_http_status

You would add something like this to your `application.properties`:

```
error.handling.full-stacktrace-http-statuses[0]=500
error.handling.full-stacktrace-http-statuses[1]=40x
error.handling.full-stacktrace-http-statuses[2]=2xx
```

And it would log full stack traces if the http response status would match. The only drawback is that it will log twice, worst case, it will log the stacktrace twice if you would have also configured logging for the Exception class that is being thrown.

So I am open for suggestions on how we can improve this to avoid the double logging. Let me know what you think!

**Comment by sverhagen** [comment]
Thank you for this solution, and for your patience. This had fallen off my radar, but I'm back.

I don't know about the logging. I am sure people manage exceptions in a million different ways. For me, generally, 5xx means that the system has gone outside of its expected behavior. Which means that I didn't really code for it. So I like to have a "broad sweep" configuration to get exceptions in that part of the spectrum logged noisely, `full-stacktrace-http-statuses: 5xx` does that for me. I guess my point is, that I have no idea of how I would this ever use in a conflicting way with `full-stacktrace-classes`.

I would use this solution, come a non-`SNAPSHOT` version. Thanks!

**Comment by wimdeblauwe** [comment]
Version 1.7.0 has been released with this included. Please give it a try and let me know if there is anything else I can help with.

**Comment by sverhagen** [comment]
Thanks, I have integrated 1.7.0 into the few services of interest, and we're gonna run with it! Much appreciated.

**Comment by sverhagen** [comment]
Just curious, what is the reason the `logException(exception);` in `ErrorHandlingControllerAdvice.handleException` can't move below the `ApiExceptionHandler`s being called?

**Comment by wimdeblauwe** [comment]
It can technically, but I like to have it first before the exception handling kicks in, just in case there is something wrong with the exception handling, there is at least the log trace.

**Comment by sverhagen** [comment]
>something wrong with the exception handling

If ^^that were to manifest itself as an exception in and of itself, perhaps a `try`/`catch` could ease your concern?

(Clearly, I got hit by the double-logging, and I had to explain a colleague why that was the case :smile:)

**Comment by wimdeblauwe** [comment]
🤔 Maybe you are right and I should rethink this to avoid the double-logging

**Comment by wimdeblauwe** [comment]
FYI: I just released version 2.0.0 that should fix the double logging. See https://github.com/wimdeblauwe/error-handling-spring-boot-starter/releases/tag/2.0.0
